src/encode.rs

```diff
@@ -0,0 +1,188 @@
+    fn cause(&self) -> Option<&error::Error> {
+        match *self {
+            Error::Bitcoin(ref e) => Some(e),
+            _ => None,
+        }
+    }
+
+    fn description(&self) -> &str {
+        "an Elements encoding error"
+    }
+}
+
+#[doc(hidden)]
+impl From<btcenc::Error> for Error {
+    fn from(e: btcenc::Error) -> Error {
+        Error::Bitcoin(e)
+    }
+}
+
+/// Data which can be encoded in a consensus-consistent way
+pub trait Encodable {
+    /// Encode an object with a well-defined format, should only ever error if
+    /// the underlying `Write` errors. Returns the number of bytes written on
+    /// success
+    fn consensus_encode<W: io::Write>(&self, e: W) -> Result<usize, Error>;
+}
+
+/// Data which can be encoded in a consensus-consistent way
+pub trait Decodable: Sized {
+    /// Decode an object with a well-defined format
+    fn consensus_decode<D: io::Read>(d: D) -> Result<Self, Error>;
+}
+
+/// Encode an object into a vector
+pub fn serialize<T: Encodable + ?Sized>(data: &T) -> Vec<u8> {
+    let mut encoder = Cursor::new(vec![]);
+    data.consensus_encode(&mut encoder).unwrap();
+    encoder.into_inner()
+}
+
+/// Encode an object into a hex-encoded string
+pub fn serialize_hex<T: Encodable + ?Sized>(data: &T) -> String {
+    ::bitcoin::hashes::hex::ToHex::to_hex(&serialize(data)[..])
+}
+
+/// Deserialize an object from a vector, will error if said deserialization
+/// doesn't consume the entire vector.
+pub fn deserialize<'a, T: Decodable>(data: &'a [u8]) -> Result<T, Error> {
+    let (rv, consumed) = deserialize_partial(data)?;
+
+    // Fail if data are not consumed entirely.
+    if consumed == data.len() {
+        Ok(rv)
+    } else {
+        Err(Error::ParseFailed("data not consumed entirely when explicitly deserializing"))
+    }
+}
+
+/// Deserialize an object from a vector, but will not report an error if said deserialization
+/// doesn't consume the entire vector.
+pub fn deserialize_partial<'a, T: Decodable>(data: &'a [u8]) -> Result<(T, usize), Error> {
+    let mut decoder = Cursor::new(data);
+    let rv = Decodable::consensus_decode(&mut decoder)?;
+    let consumed = decoder.position() as usize;
+
+    Ok((rv, consumed))
+}
+
+/// Implement Elements encodable traits for Bitcoin encodable types.
+macro_rules! impl_upstream {
+    ($type: ty) => {
+        impl Encodable for $type {
+            fn consensus_encode<W: io::Write>(&self, mut e: W) -> Result<usize, Error> {
+                Ok(btcenc::Encodable::consensus_encode(self, &mut e)?)
+            }
+        }
+
+        impl Decodable for $type {
+            fn consensus_decode<D: io::Read>(mut d: D) -> Result<Self, Error> {
+                Ok(btcenc::Decodable::consensus_decode(&mut d)?)
+            }
+        }
+    };
+}
+impl_upstream!(u8);
+impl_upstream!(u32);
+impl_upstream!(u64);
+impl_upstream!([u8; 32]);
+impl_upstream!(Vec<u8>);
+impl_upstream!(Vec<Vec<u8>>);
+impl_upstream!(btcenc::VarInt);
```

Comment:
I think we should `pub use bitcoin::VarInt` in `lib.rs`with_capacity

Reply:
Why? How about `Script`?

Reply:
An elements script is not the same as bitcoin script :). But an elements varint is a bitcoin varint.

Reply:
At some point we should have a separate `elements::script` and `elements::opcodes` modules that are independent of the Bitcoin ones, but for now (especially with our heavy Miniscript usage which is Bitcoin-only) it's fine to keep using `bitcoin::Script` in elements data structures, even though this isn't very well typed.